validator/machine_cache.go

```diff
@@ -5,70 +5,126 @@ package validator
 	cache := &MachineCache{
-		machines:          []MachineInterface{initialMachine},
-		targetNumMachines: targetNumMachines,
-		firstMachineStep:  initialMachine.GetStepCount(),
+		readyMarker: newReadyMarker(),
+		config:      config,
 	}
-	err := cache.populateInitialCache(ctx)
-	if err != nil {
-		return nil, err
-	}
-	return cache, nil
+	go func() {
+		zeroStepMachine, err := initialMachineGetter(ctx)
+		if err == nil && zeroStepMachine.GetStepCount() != 0 {
+			zeroStepMachine.Destroy()
+			err = errors.New("initialMachine not at step count 0")
+		}
+		if err != nil {
+			cache.signalReady(err)
+			return
+		}
+		zeroStepMachine.Freeze()
+		cache.zeroStepMachine = zeroStepMachine
+		cache.machines = []MachineInterface{zeroStepMachine}
+		cache.firstMachineStep = 0
+		cache.machineStepInterval = config.InitialSteps
+		err = cache.populateInitialCache(ctx, ^uint64(0))
+		if err != nil {
+			cache.signalReady(err)
+			return
+		}
+		cache.finalMachine = cache.machines[len(cache.machines)-1]
+		cache.finalMachine.Freeze()
+		cache.signalReady(nil)
+	}()
+	return cache
 }
 
-// `endSteps` should be the *total* step count at which the cache ends, not the number of steps from `initialMachine` to the end.
-// `initialMachine` may be mutated by this function.
-func NewMachineCacheWithEndSteps(ctx context.Context, initialMachine MachineInterface, targetNumMachines int, endSteps uint64) (*MachineCache, error) {
-	startSteps := initialMachine.GetStepCount()
-	if endSteps < startSteps {
-		return nil, errors.Errorf("endSteps %v before initialMachine step count %v", endSteps, startSteps)
+func (c *MachineCache) SpawnCacheWithLimits(ctx context.Context, start uint64, end uint64) *MachineCache {
+	newInterval := (start - end) / uint64(c.config.TargetMachineCount)
+	if start == c.firstMachineStep && newInterval == c.machineStepInterval {
+		return c
 	}
-	cache := &MachineCache{
-		machines:            []MachineInterface{initialMachine.CloneMachineInterface()},
-		targetNumMachines:   targetNumMachines,
-		firstMachineStep:    startSteps,
-		machineStepInterval: (endSteps - startSteps) / uint64(targetNumMachines+1),
+	newCache := &MachineCache{
+		readyMarker: newReadyMarker(),
+		config:      c.config,
 	}
-	for i := 1; i < targetNumMachines; i++ {
-		if !initialMachine.IsRunning() {
-			break
+	go func() {
+		err := c.WaitReady(ctx)
+		if err != nil {
+			newCache.signalReady(err)
+			return
 		}
-		err := initialMachine.Step(ctx, cache.machineStepInterval)
+		newCache.zeroStepMachine = c.zeroStepMachine
+		newCache.finalMachine = c.finalMachine
+		closest, err := c.getClosestMachine(start)
 		if err != nil {
-			return nil, err
+			newCache.signalReady(err)
+			return
 		}
-		cache.machines = append(cache.machines, initialMachine.CloneMachineInterface())
-	}
-	return cache, nil
+		initial := closest.CloneMachineInterface()
```

Comment:
added some fixes to last PR, including not cloning if closest is already in the right step (otherwise it's a frozen machine and must be cloned to advance)

Reply:
I will say that there is still a lot to improve on machineCache in this API.
Specifically in how PrepareRange is handled - e.g. need to destory the previous step's machines.. I think I'll add another PR after #6 to address these things. I think I tried to avoid locking and it created a worse solution.